nav2_behavior_tree/plugins/action/remove_in_collision_goals_action.cpp

```diff
@@ -63,12 +64,31 @@ BT::NodeStatus RemoveInCollisionGoals::on_completion(
     return BT::NodeStatus::FAILURE;
   }
 
+  // get the `waypoint_statuses` vector
+  std::string waypoint_statuses_id;
+  auto waypoint_statuses_id_get_res = getInput("waypoint_statuses_id", waypoint_statuses_id);
+  std::vector<nav2_msgs::msg::WaypointStatus> waypoint_statuses;
+  if (waypoint_statuses_id_get_res) {
+    waypoint_statuses =
+      config().blackboard->get<decltype(waypoint_statuses)>(waypoint_statuses_id);
+  }
+
   nav_msgs::msg::Goals valid_goal_poses;
   for (size_t i = 0; i < response->costs.size(); ++i) {
     if ((response->costs[i] == 255 && !consider_unknown_as_obstacle_) ||
       response->costs[i] < cost_threshold_)
     {
       valid_goal_poses.goals.push_back(input_goals_.goals[i]);
+    } else if (waypoint_statuses_id_get_res) {
+      using namespace nav2_util::geometry_utils;  // NOLINT
+      auto cur_waypoint_index =
+        find_goal_in_waypoint_statuses(waypoint_statuses, input_goals_.goals[i]);
+      if (cur_waypoint_index == -1) {
+        throw std::runtime_error(
```

Comment:
If all uses of -1 return type are to throw, we can throw in the function itself!